eras/alonzo/impl/src/Cardano/Ledger/Alonzo/Genesis.hs

```
@@ -145,49 +204,43 @@ instance FromCBOR AlonzoGenesis where
           <! From
 
 instance ToCBOR AlonzoGenesis where
-  toCBOR
-    AlonzoGenesis
-      { agCoinsPerUTxOWord
-      , agCostModels
-      , agPrices
-      , agMaxTxExUnits
-      , agMaxBlockExUnits
-      , agMaxValSize
-      , agCollateralPercentage
-      , agMaxCollateralInputs
-      } =
-      toEraCBOR @AlonzoEra
-        . encode
-        $ Rec AlonzoGenesis
-          !> To agCoinsPerUTxOWord
-          !> To agCostModels
-          !> To agPrices
-          !> To agMaxTxExUnits
-          !> To agMaxBlockExUnits
-          !> To agMaxValSize
-          !> To agCollateralPercentage
-          !> To agMaxCollateralInputs
+  toCBOR x@(AlonzoGenesis _ _ _ _ _ _ _ _ _) =
+    let AlonzoGenesis {..} = x
+     in toEraCBOR @AlonzoEra . encode $
+          Rec AlonzoGenesis
+            !> To agCoinsPerUTxOWord
+            !> E encodeCostModel agCostModels
+            !> To agPrices
+            !> To agMaxTxExUnits
+            !> To agMaxBlockExUnits
+            !> To agMaxValSize
+            !> To agCollateralPercentage
+            !> To agMaxCollateralInputs
+            !> To agExtraConfig
 
 instance FromJSON AlonzoGenesis where
   parseJSON = Aeson.withObject "Alonzo Genesis" $ \o -> do
     agCoinsPerUTxOWord <- o .: "lovelacePerUTxOWord"
-    agCostModels <- parseCostModels False =<< o .: "costModels"
+    cms <- parseCostModels False =<< o .: "costModels"
     agPrices <- o .: "executionPrices"
     agMaxTxExUnits <- o .: "maxTxExUnits"
     agMaxBlockExUnits <- o .: "maxBlockExUnits"
     agMaxValSize <- o .: "maxValueSize"
     agCollateralPercentage <- o .: "collateralPercentage"
     agMaxCollateralInputs <- o .: "maxCollateralInputs"
+    agExtraConfig <- o .:? "extraConfig"
+    let agCostModels = costModelsValid cms ! PlutusV1
```

For visibility: I guess we can get away with using `!` here and just error out if there is no `PlutusV1` cost model among the valid cost models, right?

---

If you are in a parser monad, it is always better to fail in that monad, instead of a runtime exception.
Moreover we should prevent any other cost models from being supplied, otherwise they will be silently ignored, which could be very surprising to downstream users:
```suggestion
    agPlutusV1CostModel <- 
      case Map.toList (costModelsValid cms) of
        [] -> fail "Expected \"PlutusV1\" cost model to be supplied"
        [(PlutusV1, pv1CostModel)] -> pure pv1CostModel
        _ -> fail $ "Only PlutusV1 CostModel is allowed in the AlonzoGenesis, but " <> List.intercalate ", " (Map.keys cms) <> " were supplied. Use \"extraConfig\" if you need to inject other cost models for testing."
```